Cython/Utility/ObjectHandling.c

```diff
@@ -2214,10 +2214,39 @@ static PyObject* __Pyx_PyObject_CallMethod1(PyObject* obj, PyObject* method_name
 
 
 /////////////// tp_new.proto ///////////////
+//@requires: ModuleSetupCode.c::GetRuntimeVersion
 
 #define __Pyx_tp_new(type_obj, args) __Pyx_tp_new_kwargs(type_obj, args, NULL)
 static CYTHON_INLINE PyObject* __Pyx_tp_new_kwargs(PyObject* type_obj, PyObject* args, PyObject* kwargs) {
-    return (PyObject*) (((PyTypeObject*)type_obj)->tp_new((PyTypeObject*)type_obj, args, kwargs));
+    newfunc tp_new = __Pyx_PyType_TryGetSlot((PyTypeObject*)type_obj, tp_new, newfunc);
+#if CYTHON_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030A0000
+    if (!tp_new) {
+        PyObject *new_func = NULL, *new_args = NULL, *obj;
+        Py_ssize_t i, nargs = PyTuple_Size(args);
+
+        if (unlikely(nargs < 0)) goto bad;
+        new_args = PyTuple_New(nargs + 1);
+        if (unlikely(!new_args)) goto bad;
+        for (i = 0; i < nargs + 1; i++) {
+            PyObject *item = (i == 0) ? type_obj : PyTuple_GetItem(args, i - 1);
+            if (unlikely(!item)) goto bad;
+            if (unlikely(PyTuple_SetItem(new_args, i, item)) < 0) goto bad;
+            Py_INCREF(item);
```

Comment:
I think it's generally bad style to incref *after* storing an external reference. I understand that it simplifies the error handling here, but it means that there's a short time interval in which the refcounts in a collection are too low. I'd rather try to avoid that – *especially* in code for the Stable ABI. There is no guarantee that future Python implementations won't replace their garbage collector with something more elaborate, for example.

And it's not like we're really risking any failures in this code…

Reply:
It doesn't actually simplify the reference counting here. `PyTuple_SetItem` steals a reference even if it fails. Therefore this code is actually wrong if it fails because it loses a reference to `item`.

Just put the incref first and it's right in both cases.

Reply:
Damn, I should have checked the implementation. Of course, it makes sense to behave like that.

Reply:
@scoder I'm not sure whether my update I captured all what you suggested, please have it a second look.